pkg/controller/helm/helm.go

```diff
@@ -0,0 +1,270 @@
+	if err != nil {
+		gl.Status.State = xpcorev1alpha1.Failed
+		gl.SetFailed(reasonCreatingApplicationProducer, err.Error())
+		return RequeueNow
+	}
+
+	app, err := ap.Produce()
+	if err != nil {
+		gl.Status.State = xpcorev1alpha1.Failed
+		gl.SetFailed(reasonProducingApplication, err.Error())
+		return RequeueNow
+	}
+
+	remote := app.DeepCopy()
+	if err := crud.CreateOrUpdate(ctx, c.kube, remote, func() error {
+		// Inside this anonymous function remote could either be unchanged (if
+		// it does not exist in the API server) or updated to reflect its
+		// current state according to the API server.
+
+		if !hasSameController(remote, app) {
+			return errors.Errorf("%s %s exists and is not controlled by %s %s",
+				xpworkloadv1alpha1.KubernetesApplicationKind, remote.GetName(),
+				v1alpha1.GitLabKind, getControllerName(app))
+		}
+
+		remote.SetLabels(app.GetLabels())
+		remote.Spec = *app.Spec.DeepCopy()
+
+		return nil
+	}); err != nil {
+		gl.Status.State = xpcorev1alpha1.Failed
+		gl.SetFailed(reasonSyncingApplication, err.Error())
+		return RequeueNow
+	}
+
+	// TODO(negz): Derive our readiness from app's status.
+	gl.Status.State = xpcorev1alpha1.Ready
+	gl.SetReady()
+
+	return RequeueIfDirty
+}
+
+func (c *localCluster) delete(ctx context.Context, gl *v1alpha1.GitLab) reconcile.Result {
+	gl.Status.SetDeleting()
+	meta.RemoveFinalizer(gl, finalizerName)
+	return RequeueIfDirty
+}
+
+// Reconciler reconciles a GitLab object
+type Reconciler struct {
+	kube  client.Client
+	local syncDeleter
+}
+
+// Reconcile reads that state of the cluster for a GitLab object and makes changes based on the state read
+// and what is in the GitLab.Spec
+// Automatically generate RBAC rules to allow the Controller to read and write KubernetesApplications
+// +kubebuilder:rbac:groups=workload.crossplane.io,resources=kubernetesapplications,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=controller.gitlab.com,resources=gitlabs,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=controller.gitlab.com,resources=gitlabs/status,verbs=get;update;patch
+func (r *Reconciler) Reconcile(req reconcile.Request) (reconcile.Result, error) {
+	log.V(logging.Debug).Info("reconciling", "kind", v1alpha1.GitLabKind, "request", req)
+
+	ctx, cancel := context.WithTimeout(context.Background(), reconcileTimeout)
+	defer cancel()
+
+	gl := &v1alpha1.GitLab{}
+	if err := r.kube.Get(ctx, req.NamespacedName, gl); err != nil {
+		if kerrors.IsNotFound(err) {
+			return RequeueIfDirty, nil
+		}
+		return RequeueIfDirty, errors.Wrapf(err, "cannot get %s %s", v1alpha1.GitLabKind, req.NamespacedName)
+	}
+
+	if gl.GetDeletionTimestamp() != nil {
+		return r.local.delete(ctx, gl), errors.Wrapf(r.kube.Update(ctx, gl), "failed to update %s %s", v1alpha1.GitLabKind, req.NamespacedName)
+	}
+
+	return r.local.sync(ctx, gl), errors.Wrapf(r.kube.Update(ctx, gl), "failed to update %s %s", v1alpha1.GitLabKind, req.NamespacedName)
```

Comment:
It is now apparent why we "need" a `combined` client since at this point in code we don't have context on what type of the update is needed (status or other). I am not sure I like this approach. I understand the desire of not "leaking" `client.Client` into "deeper" function calls, but I am afraid that is inevitable, especially if/when managing kubernetes resources other than "this" object.

Reply:
Unfortunately neither of us like the other's approach to controller patterns at the moment. For now I'm planning on getting rid of this controller and merging my Helm logic into yours. Let's circle back to our ongoing attempts to arrive at a controller pattern we all can appreciate once we've got this demo out of the way.